scalableminds/webknossos
# Dynamic bucket allocation & miscellaneous volume improvements

- increase speed of compression for volume buckets by using two workers instead of one for that job (adds a new `WorkerPool` abstraction for that)
- adds a tooltip (only visible on hover) to the save button in case there are currently buckets waiting for compression (when annotating large amounts the compression can take some seconds which might irritate the user since the progress was not really visible)
- adapts the bucket allocation logic so that there is no hard limit anymore. instead, there is an initial limit which can be exceeded if necessary. if the limit is exceeded by a factor of 2 (this corresponds to the hotfix of bumping the limit to 10000 for segmentation layers, the user is warned about that and advised to save&reload). I didn't built a shrinking logic, since I don't consider the advantages to be worth it. Happy to discuss this, though!
- fixes that `PushQueue.push` could starve when the user never does a break longer than 1 second (unrealistic, but when testing, one can easily run into this)
- refactor: rename some important variables (such as `queue`) and add some comments

### URL of deployed dev instance (used for testing):
- https://dynamicbucketallocation.webknossos.xyz

### Steps to test:
- I adapted the code to output some periodic logging about the bucket counts and use a lower maximum bucket limit, so that the critical cases can be tested easier
- open a new annotation and open the devtools
- brush large areas over multiple viewports to quickly build up a load of changed buckets
- shortly before it reaches 500, stop brushing and hit save
- hovering over the save-button should show some information about how many buckets need to be compressed
- when everything is saved, brush again (in moderate speed)
- ideally, the allocated bucket limit should not exceed 500
- now, brush very fast across all viewports and jump into new regions to force wk to allocate new buckets
- --> the bucket count should grow over 500
- continue so that the bucket count grows over 1000 --> a toast should appear
- save & wait
- reload the page --> everything should look as before


![image](https://user-images.githubusercontent.com/2486553/154708964-e5738578-7965-4fb0-8c73-748a15e0a98f.png)

![image](https://user-images.githubusercontent.com/2486553/154709059-01fd0406-a41d-4fa1-8c41-697f55ec40d1.png)


### Issues:
- contributes to #5973 

------
(Please delete unneeded items, merge only when none are left open)
